### PR TITLE
[ROCm] Adding ROm support for "concat", "pack", "unpack" and "split" ops

### PR DESCRIPTION
**Pull request by deven-amd**
This PR add ROCm support for "concat", "pack", "unpack" and "split" ops

PR #28343 is a pre-req for this PR, and hence this PR includes commits from PR #28343
Only the last three commits in this PR is exclusive to this PR, and hence should be the only one reviewed.

-------------------------------

Please ignore the cla check failure...that is getting triggered because this PR builds on top of PR #28343 (which is submitted by a different developer). Once that PR is merged, and I rebase this PR, the commit triggering the cla check failure will be gone from this commit

-------------------------------

@tatianashp , @whchung : just FYI

### CODE REVIEW

**Comment by timshen91** [comment]
@chsigg , would you be able to take a look at this and PR 28450 as they are TF implementation changes?

**Comment by deven-amd** [comment]
@chsigg

rebased this PR now that the updates related to Cuda* to Gpu* renaming are merged into master.

This PR is now trivial, please review and merge.

thanks

**Review by whchung — DISMISSED** [review]

**Comment by deven-amd** [comment]
The failures in the `LInux GPU` CI run are due to the fact that PR #28834 is not yet merged.

PR #28834 renames the `cuda_helper` namespace to `gpu_helper` (an alias `cuda_helper` is also introduced for backward compatbility).  

This PR updates the the `cuda_helper` reference to `gpu_helper` (since the `cuda_helper` name is not visible in ROCm builds). Please re-run CI after PR #28834 is merged, that should get rid of the errors.

thanks

File: tensorflow/core/kernels/concat_lib_gpu_impl.cu.cc [hunk 1]
```diff
@@ -156,18 +161,19 @@ void ConcatGPUImpl(const Eigen::GpuDevice& gpu_device,
...
-                                   output->dimension(1), output->data()));
-    else
-      TF_CHECK_OK(CudaLaunchKernel(concat_variable_kernel<T, IntType, false>,
-                                   config.block_count, config.thread_per_block,
-                                   0, gpu_device.stream(), input_ptrs,
-                                   output_scan, output->dimension(0),
-                                   output->dimension(1), output->data()));
+    if (smem_usage < smem_max && smem_usage < kMaxSmemBytesPerformance) {
+      TF_CHECK_OK(GpuLaunchKernel(
+          (concat_variable_kernel<T, IntType, true>), dim3(config.block_count),
```
Comment (chsigg) [inline tensorflow/core/kernels/concat_lib_gpu_impl.cu.cc]: Remove extra parenthesis. Here and below.

File: tensorflow/core/kernels/split_lib_gpu.cu.cc [hunk 2]
```diff
@@ -124,7 +124,11 @@ __global__ void split_v_kernel(const T* input_ptr,
   int num_outputs = output_ptr_data.size;
 
   // verbose declaration needed due to template
+#if GOOGLE_CUDA
```
Comment (chsigg) [inline tensorflow/core/kernels/split_lib_gpu.cu.cc]: Have you considered introducing a wrapper for this?

File: tensorflow/core/kernels/split_lib_gpu.cu.cc [hunk 3]
```diff
@@ -217,10 +221,10 @@ void SplitVOpGPULaunch<T, IntType>::Run(
     GpuLaunchConfig config =
         GetGpuLaunchConfig(total_rows * total_cols, gpu_device);
 
-    TF_CHECK_OK(CudaLaunchKernel(SplitVOpKernel_fixed<T>, config.block_count,
-                                 config.thread_per_block, 0,
-                                 gpu_device.stream(), input_ptr, total_rows,
-                                 total_cols, output_ptr_data));
+    TF_CHECK_OK(
+        GpuLaunchKernel(SplitVOpKernel_fixed<T>, dim3(config.block_count),
```
Comment (chsigg) [inline tensorflow/core/kernels/split_lib_gpu.cu.cc]: Does ROCm's dim3 not implicitly construct from int?

File: tensorflow/core/kernels/concat_lib_gpu_impl.cu.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/core/kernels/concat_lib_gpu_impl.cu.cc]: let me do that and push out a rebase

File: tensorflow/core/kernels/split_lib_gpu.cu.cc (hunk 3 above)
Comment (deven-amd) [reply tensorflow/core/kernels/split_lib_gpu.cu.cc]: it does...the dim3 call does not need to be there...will remove it along with the extra parens

File: tensorflow/core/kernels/split_lib_gpu.cu.cc (hunk 2 above)
Comment (deven-amd) [reply tensorflow/core/kernels/split_lib_gpu.cu.cc]: I had not until now.

would something the following lines work

```
<in tensorflow/core/gpu_kernel_helper.h>

#if GOOGLE_CUDA
  #define GPU_DYNAMIC_SHARED_MEM_DECL(ALIGN, TYPE, NAME)     extern __shared__ __align__(ALIGN) TYPE NAME[]
#elif TENSORFLOW_USE_ROCM
  #define GPU_DYNAMIC_SHARED_MEM_DECL(ALIGN, TYPE, NAME)     HIP_DYNAMIC_SHARED(TYPE, NAME)
#endif 


<and then in the file where we need the shared memory decl, like right here>

GPU_DYNAMIC_SHARED_MEM_DECL(sizeof(T), unsigned char, smem);

```

There are 8 instances of HIP_DYNAMIC_SHARED in our fork...so this should not be a major change.

Assuming the above change is the route we want to go, can we track it in a different PR, and leave the change here as is?

thanks

Comment (deven-amd) [reply tensorflow/core/kernels/split_lib_gpu.cu.cc]: @chsigg , gentle ping ... let me know if you are okay with the proposal above, and whether we can address it in a separate PR.

**Comment by chsigg** [comment]
Yes, we can address the dynamic shared memory in a separate change.

File: tensorflow/core/kernels/split_lib_gpu.cu.cc (hunk 2 above)
Comment (deven-amd) [reply tensorflow/core/kernels/split_lib_gpu.cu.cc]: cool. expect a PR for it sometime next week.

also this PR should be good to go now.

thanks again

**Review by chsigg — APPROVED** [review]

**Pull request merged**